src/main/java/com/fasterxml/jackson/databind/deser/BeanDeserializer.java

```diff
@@ -296,9 +296,10 @@ private final Object vanillaDeserialize(JsonParser p,
         throws IOException
     {
         final Object bean = _valueInstantiator.createUsingDefault(ctxt);
-        // [databind#631]: Assign current value, to be accessible by custom serializers
-        p.setCurrentValue(bean);
```

Comment:
This is the actual fix: for "empty" Objects at this point parser points to `JsonToken.END_OBJECT` and hence parent context -- we do not want to overwrite it.